llmfoundry/eval/datasets/in_context_learning_evaluation.py

```diff
@@ -1323,108 +1513,46 @@ def build_icl_dataloader(
-            hf_loading_vars=hf_loading_vars,
-            hf_parsing_map=hf_parsing_map,
-            generation_kwargs=generation_kwargs,
-        )
-        batch_size = max(dataset.num_choices, batch_size)
-        effective_batchsize = batch_size // dataset.num_choices
-    elif icl_task_type == 'schema':
-        dataset = InContextLearningSchemaTaskDataset(
-            dataset_uri=dataset_uri,
-            tokenizer=tokenizer,
-            max_seq_len=max_seq_len,
-            pad_tok_id=pad_tok_id,
-            num_fewshot=num_fewshot,
-            prompt_string=prompt_string,
-            example_delimiter=example_delimiter,
-            continuation_delimiter=continuation_delimiter,
-            destination_path=destination_path,
-            prelimiter=prelimiter,
-            fewshot_random_seed=fewshot_random_seed,
-            hf_loading_vars=hf_loading_vars,
-            hf_parsing_map=hf_parsing_map,
-            generation_kwargs=generation_kwargs,
-        )
-        batch_size = max(dataset.num_choices, batch_size)
-        effective_batchsize = batch_size // dataset.num_choices
-    elif icl_task_type == 'language_modeling':
-        dataset = InContextLearningLMTaskDataset(
-            dataset_uri=dataset_uri,
-            tokenizer=tokenizer,
-            max_seq_len=max_seq_len,
-            pad_tok_id=pad_tok_id,
-            num_fewshot=num_fewshot,
-            prompt_string=prompt_string,
-            example_delimiter=example_delimiter,
-            continuation_delimiter=continuation_delimiter,
-            destination_path=destination_path,
-            prelimiter=prelimiter,
-            fewshot_random_seed=fewshot_random_seed,
-            hf_loading_vars=hf_loading_vars,
-            hf_parsing_map=hf_parsing_map,
-            generation_kwargs=generation_kwargs,
-        )
-        effective_batchsize = batch_size
-    elif icl_task_type == 'generation_task_with_answers':
-        dataset = InContextLearningGenerationTaskWithAnswersDataset(
-            dataset_uri=dataset_uri,
-            tokenizer=tokenizer,
-            max_seq_len=max_seq_len,
-            pad_tok_id=pad_tok_id,
-            num_fewshot=num_fewshot,
-            prompt_string=prompt_string,
-            example_delimiter=example_delimiter,
-            continuation_delimiter=continuation_delimiter,
-            destination_path=destination_path,
-            prelimiter=prelimiter,
-            fewshot_random_seed=fewshot_random_seed,
-            hf_loading_vars=hf_loading_vars,
-            hf_parsing_map=hf_parsing_map,
-            cot_delimiter=cot_delimiter,
-            early_stopping_criteria=early_stopping_criteria,
-            do_normalization=do_normalization,
-            generation_kwargs=generation_kwargs,
-        )
-        effective_batchsize = batch_size
-    else:
-        raise Exception(f'Unrecognized ICL task type: {icl_task_type}')
-
+    name = icl_task_type
+    kwargs: Dict[str, Any] = {
```

Comment:
I think we can get rid of this hardcoding of the kwargs, and remove the *args, **kwargs from the Dataset classes, by:
(1) Move the defaulters [here](https://github.com/mosaicml/llm-foundry/blob/5571101a50804406ef0fe23e7ea6795b3c4a1bcb/llmfoundry/utils/builders.py#L548-L563) into the Dataset class constructors themselves, so they that aren't hardcoded in some other place.
(2) just pass kwargs, all the way through until the construction from the registry

Does that make sense?

Reply:
Points 1,2 make sense to me. But will we not need `*args, **kwargs` in the Dataset classes to catch any extra parameters which get sent from the registry constructor, since the arguments vary slightly for every icl dataset type.

Reply:
Actually might be alright

Reply:
Yeah, my thought is that the only reason the *args/**kwargs exists is because the defaults are set independent of dataset type, and so some of those defaults are passed to constructors that don't need them. If the defaults move into the constructors, each dataset should only be receiving args that it actually uses.

Reply:
So are you suggesting constructing the kwargs for this call https://github.com/mosaicml/llm-foundry/blob/5571101a50804406ef0fe23e7ea6795b3c4a1bcb/llmfoundry/utils/builders.py#L601 so that based on the dataset type we choose which fields to include (for eg, cot_delimiter will get included only in a particular dataset). 

With kwargs in the dataset constructors, I can use a single call like:
```
            icl_constructor_kwargs: Dict[str, Any] = {
                'dataset_uri': icl_cfg['dataset_uri'],
                'tokenizer': tokenizer,
                'max_seq_len': icl_cfg['max_seq_len'],
                'pad_tok_id': pad_tok_id,
                'num_fewshot': num_fewshot,
                'prompt_string': icl_cfg['prompt_string'],
                'example_delimiter': icl_cfg['example_delimiter'],
                'continuation_delimiter': icl_cfg['continuation_delimiter'],
                'question_prelimiter': icl_cfg.get('question_prelimiter', ''),
                'fewshot_random_seed': icl_cfg['fewshot_random_seed'],
                'cot_delimiter': icl_cfg.get('cot_delimiter', ''),
                'early_stopping_criteria': early_stopping_criteria,
                'do_normalization': icl_cfg.get('do_normalization', True),
                'generation_kwargs': icl_cfg.get('generation_kwargs', {}),
            }

            dataloaders = get_icl_task_dataloader(
                icl_task_type=icl_cfg['icl_task_type'],
                dataset_uri=icl_cfg['dataset_uri'],
                tokenizer=tokenizer,
                batch_size=icl_cfg['batch_size'],
                hf_loading_vars=hf_loading_vars,
                hf_parsing_map=hf_parsing_map,
                has_categories=icl_cfg.get('has_categories', False),
                destination_path=destination_path,
                kwargs=icl_constructor_kwargs
            )
```

Reply:
The problem with kwargs in the dataset constructors that get ignored is that it can silently ignore typos/code errors. I am suggesting removing all of this code (https://github.com/mosaicml/llm-foundry/blob/5571101a50804406ef0fe23e7ea6795b3c4a1bcb/llmfoundry/utils/builders.py#L548-L569), so that there are no dataset specific args that are getting passed into other constructors that don't use them. These defaults can be moved into the relevant dataset constructor. Then icl_cfg can be passed along to [here](https://github.com/mosaicml/llm-foundry/blob/main/llmfoundry/eval/datasets/in_context_learning_evaluation.py#L1511) as arbitrary kwargs before being **ed into the registried constructor. Things that are not in `icl_cfg` (e.g. tokenizer) can be passed directly by named arg since all the classes accept these in their constructor.

Happy to help with implementation too if I am not being clear. Feel free to message me on slack as well.